{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chain ID not provided for PDB 7VV3.\n",
      "Chain ID not provided for PDB 7VUY.\n",
      "Chain ID not provided for PDB 7VUZ.\n",
      "Removed the following PDB codes: {'8ZPS', '8F2A', '7F8V', '7TYX', '7TZF', '7TYL', '7SK6', '7FIY', '7TYI', '7SK5', '8F2B', '7P02', '7X1U', '7VDL', '7TYH', '8K9L', '7RMH', '7XBD', '8DWC'}\n",
      "PDB file 8JPS.pdb downloaded successfully.\n",
      "PDB file 7SK4.pdb downloaded successfully.\n",
      "PDB file 8GY7.pdb downloaded successfully.\n",
      "PDB file 7EIB.pdb downloaded successfully.\n",
      "PDB file 7F6I.pdb downloaded successfully.\n",
      "PDB file 8IA8.pdb downloaded successfully.\n",
      "PDB file 8I95.pdb downloaded successfully.\n",
      "PDB file 8HK2.pdb downloaded successfully.\n",
      "PDB file 8F0K.pdb downloaded successfully.\n",
      "PDB file 9AUC.pdb downloaded successfully.\n",
      "PDB file 7TYO.pdb downloaded successfully.\n",
      "PDB file 7VL9.pdb downloaded successfully.\n",
      "PDB file 7XA3.pdb downloaded successfully.\n",
      "PDB file 7YKD.pdb downloaded successfully.\n",
      "PDB file 8JJP.pdb downloaded successfully.\n",
      "PDB file 8XGM.pdb downloaded successfully.\n",
      "PDB file 8IC0.pdb downloaded successfully.\n",
      "PDB file 8HNK.pdb downloaded successfully.\n",
      "PDB file 8HCQ.pdb downloaded successfully.\n",
      "PDB file 7VFX.pdb downloaded successfully.\n",
      "PDB file 7T6T.pdb downloaded successfully.\n",
      "PDB file 7WQ3.pdb downloaded successfully.\n",
      "PDB file 7WQ4.pdb downloaded successfully.\n",
      "PDB file 8IA7.pdb downloaded successfully.\n",
      "PDB file 7XOW.pdb downloaded successfully.\n",
      "PDB file 7RBT.pdb downloaded successfully.\n",
      "PDB file 7FIN.pdb downloaded successfully.\n",
      "PDB file 8YW4.pdb downloaded successfully.\n",
      "PDB file 7W40.pdb downloaded successfully.\n",
      "PDB file 7W3Z.pdb downloaded successfully.\n",
      "PDB file 8XGO.pdb downloaded successfully.\n",
      "PDB file 8XGS.pdb downloaded successfully.\n",
      "PDB file 8WVY.pdb downloaded successfully.\n",
      "PDB file 8IOC.pdb downloaded successfully.\n",
      "PDB file 8IOD.pdb downloaded successfully.\n",
      "PDB file 8INR.pdb downloaded successfully.\n",
      "PDB file 8WSS.pdb downloaded successfully.\n",
      "PDB file 8WST.pdb downloaded successfully.\n",
      "PDB file 8JGF.pdb downloaded successfully.\n",
      "PDB file 8JGB.pdb downloaded successfully.\n",
      "PDB file 7S8L.pdb downloaded successfully.\n",
      "PDB file 7VDM.pdb downloaded successfully.\n",
      "PDB file 7VV0.pdb downloaded successfully.\n",
      "PDB file 8IBV.pdb downloaded successfully.\n",
      "PDB file 8U26.pdb downloaded successfully.\n",
      "PDB file 7XWO.pdb downloaded successfully.\n",
      "Failed to download PDB file 8JBG. Status code: 404\n",
      "Failed to download PDB file 8JBH. Status code: 404\n",
      "PDB file 8JBF.pdb downloaded successfully.\n",
      "PDB file 8H0P.pdb downloaded successfully.\n",
      "PDB file 7W56.pdb downloaded successfully.\n",
      "PDB file 7W53.pdb downloaded successfully.\n",
      "PDB file 7W55.pdb downloaded successfully.\n",
      "PDB file 7W57.pdb downloaded successfully.\n",
      "PDB file 7VGX.pdb downloaded successfully.\n",
      "PDB file 7YON.pdb downloaded successfully.\n",
      "PDB file 7YOO.pdb downloaded successfully.\n",
      "PDB file 7X9C.pdb downloaded successfully.\n",
      "PDB file 8FEG.pdb downloaded successfully.\n",
      "PDB file 8F7W.pdb downloaded successfully.\n",
      "PDB file 8K9K.pdb downloaded successfully.\n",
      "PDB file 8F7Q.pdb downloaded successfully.\n",
      "PDB file 8F7R.pdb downloaded successfully.\n",
      "PDB file 8F7X.pdb downloaded successfully.\n",
      "PDB file 7RYC.pdb downloaded successfully.\n",
      "PDB file 8ZPT.pdb downloaded successfully.\n",
      "PDB file 8WZ2.pdb downloaded successfully.\n",
      "PDB file 8ZH8.pdb downloaded successfully.\n",
      "Failed to download PDB file 8XIP. Status code: 404\n",
      "PDB file 7T10.pdb downloaded successfully.\n",
      "PDB file 7XNA.pdb downloaded successfully.\n",
      "PDB file 7T11.pdb downloaded successfully.\n",
      "PDB file 7XAV.pdb downloaded successfully.\n",
      "Failed to download PDB file 8XIR. Status code: 404\n",
      "Failed to download PDB file 8XIQ. Status code: 404\n",
      "PDB file 7XMS.pdb downloaded successfully.\n",
      "PDB file 8X8L.pdb downloaded successfully.\n",
      "PDB file 8X8N.pdb downloaded successfully.\n",
      "PDB file 8ZCJ.pdb downloaded successfully.\n",
      "PDB file 7VQX.pdb downloaded successfully.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests \n",
    "import ast\n",
    "import os\n",
    "\n",
    "def get_sequence_from_pdb_fasta(pdb_id, chain_id):\n",
    "    '''\n",
    "    Helper function to get a fasta sequence from RCSB PDB using the PDB identifier and the chain ID\n",
    "    '''\n",
    "    from Bio import SeqIO\n",
    "    import requests\n",
    "    from io import StringIO\n",
    "\n",
    "    # Link to the fasta file on RCSB PDB\n",
    "    url = f\"https://www.rcsb.org/fasta/entry/{pdb_id}/display\"\n",
    "    fasta = requests.get(url).text\n",
    "\n",
    "    # If chain_id is not provided, return None\n",
    "    if chain_id.strip() == \"\":\n",
    "        print(f\"Chain ID not provided for PDB {pdb_id}.\")\n",
    "        return None\n",
    "\n",
    "    # Return the requested sequence - try first the author chain ID, then the PDB chain ID\n",
    "    fasta_iterator = SeqIO.parse(StringIO(fasta), \"fasta\")\n",
    "    for seq in fasta_iterator:\n",
    "        if f\"[auth {chain_id}]\" in seq.description:\n",
    "            return str(seq.seq)\n",
    "\n",
    "    fasta_iterator = SeqIO.parse(StringIO(fasta), \"fasta\")\n",
    "    for seq in fasta_iterator:\n",
    "        if f\"Chain {chain_id}\" in seq.description:\n",
    "            return str(seq.seq)\n",
    "\n",
    "    # If the sequence is not found, return None\n",
    "    return None\n",
    "\n",
    "def download_structure(pdb_id, output_folder, overwrite = False, type = \"pdb\"):\n",
    "    '''\n",
    "    Function to download PDB files using the RCSB PDB API given a PDB identifier \n",
    "    and output folder name. If the PDB file already exists in the output folder,\n",
    "    the function will not download it again unless overwrite is set to True.\n",
    "    Returns the path of the downloaded PDB file.\n",
    "    '''\n",
    "\n",
    "    # Check if the PDB file already exists in the output folder\n",
    "    if os.path.exists(f\"{output_folder}/{pdb_id}.pdb\") and not overwrite:\n",
    "        print(f\"PDB file {pdb_id}.pdb already exists in the output folder.\")\n",
    "        return f\"{output_folder}/{pdb_id}.pdb\"\n",
    "\n",
    "    # Make sure output folder exists and create it if it doesn't exist\n",
    "    if not os.path.exists(output_folder):\n",
    "        os.makedirs(output_folder)\n",
    "\n",
    "    # URL for the PDB file\n",
    "    pdb_url = f\"https://files.rcsb.org/download/{pdb_id}.{type}\"\n",
    "\n",
    "    # Send an HTTP GET request to the PDB URL\n",
    "    response = requests.get(pdb_url)\n",
    "\n",
    "    # Check if the request was successful (status code 200)\n",
    "    if response.status_code == 200:\n",
    "        with open(f\"{output_folder}/{pdb_id}.pdb\", \"wb\") as pdb_file:\n",
    "            pdb_file.write(response.content)\n",
    "        print(f\"PDB file {pdb_id}.pdb downloaded successfully.\")\n",
    "        return f\"{output_folder}/{pdb_id}.pdb\"\n",
    "    else:\n",
    "        print(f\"Failed to download PDB file {pdb_id}. Status code: {response.status_code}\")\n",
    "        return \"\"\n",
    "\n",
    "def sequence_similarity(seq1, seq2):\n",
    "    from Bio import pairwise2\n",
    "    \"\"\"Calculate the sequence similarity by finding the best alignment and considering the number of identical matches.\"\"\"\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "    best_alignment = alignments[0]\n",
    "    max_length = max(len(seq1), len(seq2))\n",
    "    similarity = best_alignment.score / max_length\n",
    "    return similarity\n",
    "\n",
    "def get_receptor_sequence(uniprot_id):\n",
    "    '''\n",
    "    Function to get fasta sequence of the receptor without its signal peptide using UniProt API.\n",
    "    Returns the fasta sequence and the description of the entry given a UniProt ID.\n",
    "    '''\n",
    "\n",
    "    # Get features of the receptor from UniProt\n",
    "    url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.json\"   \n",
    "    response = requests.get(url)\n",
    "\n",
    "    if response.status_code == 200:\n",
    "\n",
    "        # Parse the JSON response\n",
    "        data = response.json()\n",
    "\n",
    "        try:\n",
    "            # Get the full sequence\n",
    "            sequence = data[\"sequence\"][\"value\"]\n",
    "        except KeyError:\n",
    "            return None, None\n",
    "\n",
    "        try:\n",
    "            # Extract feature information\n",
    "            features = data[\"features\"]\n",
    "        except KeyError:\n",
    "            return None, None\n",
    "\n",
    "        # Get the start and end positions of the chain and signal peptide\n",
    "        for feature in features:\n",
    "            if feature[\"type\"] == \"Chain\":\n",
    "                chain_start = feature[\"location\"][\"start\"][\"value\"]\n",
    "                chain_end = feature[\"location\"][\"end\"][\"value\"]\n",
    "                return feature[\"description\"], sequence[chain_start-1:chain_end]\n",
    "    else:\n",
    "        print(f\"Error: Unable to retrieve data from UniProt using ID {uniprot_id}. Status code {response.status_code}\")\n",
    "        return None, None\n",
    "    \n",
    "def get_receptor_list(receptors = []):\n",
    "    \"\"\"\n",
    "    Helper function to get a list of GPCRdb receptors with their UniProt accession numbers.\n",
    "    \"\"\"\n",
    "    # Add UniProt accession numbers of the receptors to the dataframe\n",
    "    response = requests.get(\"https://gpcrdb.org/services/receptorlist/\")\n",
    "    receptor_list = response.json()\n",
    "    receptor_list = pd.DataFrame.from_dict(receptor_list)\n",
    "    if len(receptors) != 0:\n",
    "        receptor_list = receptor_list[receptor_list[\"entry_name\"].isin(receptors)]\n",
    "    receptor_list = receptor_list[[\"entry_name\", \"accession\"]]\n",
    "    receptor_list.columns = [\"protein\", \"accession\"]\n",
    "    receptor_list.reset_index(inplace=True, drop=True)\n",
    "    return receptor_list\n",
    "\n",
    "def get_peptide_ligands(pdb_codes = []):\n",
    "    \"\"\"\n",
    "    Helper function to get a dataframe of peptide ligands with PDB structures from GPCRdb\n",
    "    \"\"\"\n",
    "\n",
    "    # Get peptide ligands for the pdb files that have been found\n",
    "    response = requests.get(\"https://gpcrdb.org/services/ligands/peptides/\")\n",
    "    peptide_ligands = response.json()\n",
    "    peptide_ligands = pd.DataFrame.from_dict(peptide_ligands)\n",
    "    if len(pdb_codes) != 0:\n",
    "        peptide_ligands = peptide_ligands[peptide_ligands[\"PDB\"].isin(pdb_codes)]\n",
    "    peptide_ligands.drop(columns=[\"Sequence length\", \"Family\", \"GPCR Class\"], inplace=True)\n",
    "    peptide_ligands.columns = [\"name\", \"sequence\", \"pdb_code\", \"chain\", \"protein\"]\n",
    "\n",
    "    return peptide_ligands\n",
    "\n",
    "def get_structure_list(species = \"Homo sapiens\"):\n",
    "    \"\"\"\n",
    "    Helper function to get a list of PDB structures from GPCRdb\n",
    "    \"\"\"\n",
    "    # Get structure information from GPCRdb for the pdb files included\n",
    "    response = requests.get('https://gpcrdb.org/services/structure/')\n",
    "    structures = response.json()\n",
    "    structures = [structure for structure in structures if structure[\"species\"] == species]\n",
    "    return structures\n",
    "\n",
    "def get_peptide_interaction(pdb_code):\n",
    "    \"\"\"\n",
    "    Helper function to get peptide interaction data of a PDB structure from GPCRdb\n",
    "    \"\"\"\n",
    "    # Get structure information from GPCRdb for the pdb files included\n",
    "    response = requests.get(f'https://gpcrdb.org/services/structure/{pdb_code}/peptideinteraction/')\n",
    "    response = response.json()\n",
    "    # Make into a dataframe\n",
    "    response_df = pd.DataFrame.from_dict(response)\n",
    "    return response_df\n",
    "\n",
    "file_dir = os.getcwd()\n",
    "repo_name = \"GPCR_peptide_benchmarking\"\n",
    "index = file_dir.find(repo_name)\n",
    "repo_dir = file_dir[:index + len(repo_name)]\n",
    "filepath = f\"{repo_dir}/classifier_benchmark_data/output/3f_known_structures_summary_2021-09-30.csv\"\n",
    "outfile = \"3f_known_structures_benchmark_2021-09-30_updated.csv\"\n",
    "outdir = f\"{repo_dir}/structure_benchmark_data/updated_data_Feb2025\"\n",
    "os.makedirs(f\"{outdir}\", exist_ok=True)\n",
    "\n",
    "# Get receptors that have peptide complexes with more recent structures\n",
    "df = pd.read_csv(filepath)\n",
    "df = df[df[\"has_peptide_complex_before_cutoff\"] == False]\n",
    "df = df[df[\"has_peptide_complex\"] == True]\n",
    "\n",
    "# Get the pdb codes for each receptor and make a dictionary\n",
    "pdbs = {}\n",
    "for i, row in df.iterrows():\n",
    "    pdb_codes = ast.literal_eval(row[\"Peptide complex PDBs after\"])\n",
    "    pdb_codes = [x[0] for x in pdb_codes]\n",
    "    pdbs[row[\"Target GPCRdb ID\"]] = pdb_codes\n",
    "\n",
    "# Make a dataframe from dict and make multiple rows for each key if there are multiple values\n",
    "pdb_code_df = pd.DataFrame.from_dict(pdbs, orient=\"index\")\n",
    "pdb_code_df = pdb_code_df.stack().reset_index()\n",
    "pdb_code_df = pdb_code_df.drop(columns=[\"level_1\"])\n",
    "pdb_code_df.columns = [\"Target GPCRdb ID\", \"PDB code\"]\n",
    "pdb_code_df = pdb_code_df.merge(df[[\"Target GPCRdb ID\", \"family\"]], on=\"Target GPCRdb ID\", how=\"left\")\n",
    "\n",
    "# Get unique pdb codes\n",
    "pdb_codes = pdb_code_df[\"PDB code\"].unique()\n",
    "\n",
    "# Get peptide ligands and list of structures from GPCRdb\n",
    "peptide_ligands = get_peptide_ligands(pdb_codes)\n",
    "structures = get_structure_list()\n",
    "\n",
    "# Filter out the structures whose ligand is not a peptide\n",
    "receptors = []\n",
    "for structure in structures:\n",
    "    for ligand in structure[\"ligands\"]:\n",
    "        if ligand[\"type\"] in [\"peptide\", \"protein\"] and structure[\"pdb_code\"] in pdb_codes:\n",
    "            keys = [k for k in structure.keys() if k not in [\"ligands\", \"signalling_protein\"]]\n",
    "            line = [structure[k] for k in keys]\n",
    "            line.extend([ligand[\"name\"], ligand[\"function\"]])\n",
    "            receptors.append(line)\n",
    "\n",
    "# Create a dataframe from the list of peptide receptors and set column names\n",
    "keys.extend([\"name\", \"function\"])\n",
    "receptors = pd.DataFrame(receptors)\n",
    "receptors.columns = keys\n",
    "\n",
    "# Merge peptide_ligands and receptors dataframes\n",
    "benchmark_set = pd.merge(receptors, peptide_ligands, on = [\"pdb_code\", \"protein\", \"name\"], how=\"left\")\n",
    "benchmark_set.rename(\n",
    "    columns={\n",
    "        'preferred_chain': 'receptor_chain',\n",
    "        'sequence': 'peptide_sequence'\n",
    "    }, inplace=True\n",
    ")\n",
    "\n",
    "# Get the UniProt accession numbers of the receptors from GPCRdb\n",
    "receptor_list = get_receptor_list(receptors[\"protein\"].unique())\n",
    "\n",
    "# Merge with receptors dataframe to add UniProt accession numbers\n",
    "benchmark_set = pd.merge(benchmark_set, receptor_list, on = [\"protein\"], how=\"left\")\n",
    "\n",
    "peptide_interactions = []\n",
    "for pdb_code in benchmark_set[\"pdb_code\"].unique():\n",
    "    peptide_interaction = get_peptide_interaction(pdb_code)\n",
    "    peptide_interactions.append(peptide_interaction)\n",
    "\n",
    "# Drop duplicates based on pdb_code and ligand_chain\n",
    "peptide_interactions_df = pd.concat(peptide_interactions)\n",
    "peptide_interactions_df = peptide_interactions_df[[\"pdb_code\", \"ligand_chain\"]].drop_duplicates()\n",
    "\n",
    "# If \"chain\" is NaN, replace it with the value from \"peptide_interactions_df\"\n",
    "benchmark_set = benchmark_set.merge(peptide_interactions_df, on=\"pdb_code\", how=\"left\")\n",
    "benchmark_set[\"chain\"] = benchmark_set[\"chain\"].fillna(benchmark_set[\"ligand_chain\"])\n",
    "benchmark_set.drop(columns=[\"ligand_chain\"], inplace=True)\n",
    "\n",
    "# Rename accession column to uniprot_id, and sort by protein and name\n",
    "benchmark_set = benchmark_set.rename(columns={\"accession\": \"receptor_uniprot_id\"})\n",
    "benchmark_set = benchmark_set.sort_values(by=[\"protein\", \"name\"])\n",
    "benchmark_set.reset_index(drop=True, inplace=True)\n",
    "benchmark_set = benchmark_set.rename(columns={\"chain\": \"ligand_chain\"})\n",
    "benchmark_set.drop(columns=[\"family\"], inplace=True)\n",
    "\n",
    "# Merge pdb_code_df with benchmark_set\n",
    "pdb_code_df.rename(columns={\"PDB code\": \"pdb_code\"}, inplace=True)\n",
    "benchmark_set = benchmark_set.merge(pdb_code_df, on=\"pdb_code\")\n",
    "benchmark_set[\"receptor_pdb_seq\"] = None\n",
    "benchmark_set[\"ligand_pdb_seq\"] = None\n",
    "\n",
    "# Drop rows where ligand_chain is NaN\n",
    "benchmark_set = benchmark_set[~benchmark_set[\"ligand_chain\"].isna()]\n",
    "\n",
    "# Get the amino acid sequences from PDB files\n",
    "for index, row in benchmark_set.iterrows():\n",
    "    receptor_pdb_seq = get_sequence_from_pdb_fasta(row[\"pdb_code\"], row[\"receptor_chain\"])\n",
    "    ligand_pdb_seq = get_sequence_from_pdb_fasta(row[\"pdb_code\"], row[\"ligand_chain\"])\n",
    "    benchmark_set.at[index, \"receptor_pdb_seq\"] = receptor_pdb_seq\n",
    "    benchmark_set.at[index, \"ligand_pdb_seq\"] = ligand_pdb_seq\n",
    "\n",
    "# Reorder and drop unnecessary columns\n",
    "col_order = {\n",
    "    'pdb_code': 'pdb',\n",
    "    'protein': 'receptor', \n",
    "    'receptor_uniprot_id': 'uniprot_id',\n",
    "    'class': 'class',\n",
    "    'family': 'family',\n",
    "    'resolution': 'resolution',\n",
    "    'type': 'type',\n",
    "    'state': 'state',\n",
    "    'name': 'ligand_name',\n",
    "    'ligand_chain': 'ligand_chain',\n",
    "    'ligand_pdb_seq': 'ligand_pdb_seq',\n",
    "    'receptor_chain': 'receptor_chain',\n",
    "    'receptor_pdb_seq': 'receptor_pdb_seq',\n",
    "    'publication': 'publication', \n",
    "    'publication_date': 'publication_date',\n",
    "}\n",
    "benchmark_set = benchmark_set[col_order.keys()]\n",
    "benchmark_set = benchmark_set.rename(columns=col_order)\n",
    "benchmark_set = benchmark_set[col_order.values()]\n",
    "\n",
    "# Drop rows where ligand_pdb_seq is none\n",
    "benchmark_set = benchmark_set[~benchmark_set[\"ligand_pdb_seq\"].isna()]\n",
    "\n",
    "# Remove pdb codes that have multiple ligands\n",
    "duplicated_pdbs = benchmark_set[benchmark_set.duplicated(subset=[\"pdb\"])]\n",
    "benchmark_set = benchmark_set[~benchmark_set[\"pdb\"].isin(duplicated_pdbs[\"pdb\"])]\n",
    "\n",
    "# Remove 7SK7 and 7SK8 from the benchmark set - there is another structure without an additional small molecule ligand\n",
    "benchmark_set = benchmark_set[benchmark_set[\"pdb\"] != \"7SK7\"]\n",
    "benchmark_set = benchmark_set[benchmark_set[\"pdb\"] != \"7SK8\"]\n",
    "\n",
    "# Drop rows where ligand_chain and receptor_chain are the same\n",
    "benchmark_set = benchmark_set[benchmark_set[\"ligand_chain\"] != benchmark_set[\"receptor_chain\"]]\n",
    "\n",
    "# Drop duplicates based on ligand_pdb_seq and receptor_pdb_seq\n",
    "benchmark_set = benchmark_set.sort_values(by=[\"receptor\", \"resolution\"])\n",
    "pdbs_before = benchmark_set[\"pdb\"]\n",
    "benchmark_set = benchmark_set.drop_duplicates(subset=[\"ligand_pdb_seq\", \"receptor_pdb_seq\"])\n",
    "pdbs_after = benchmark_set[\"pdb\"]\n",
    "\n",
    "# Check if any pdb codes were removed\n",
    "removed_pdbs = set(pdbs_before) - set(pdbs_after)\n",
    "if removed_pdbs:\n",
    "    print(f\"Removed the following PDB codes: {removed_pdbs}\")\n",
    "\n",
    "# If ligand_pdb_seq ends with X, remove it (C-terminal amidation)\n",
    "benchmark_set[\"ligand_pdb_seq\"] = benchmark_set[\"ligand_pdb_seq\"].str.rstrip(\"X\")\n",
    "\n",
    "# Replace X with A for 7W40 ligand_pdb_seq\n",
    "benchmark_set.loc[benchmark_set[\"pdb\"] == \"7W40\", \"ligand_pdb_seq\"] = benchmark_set.loc[benchmark_set[\"pdb\"] == \"7W40\", \"ligand_pdb_seq\"].str.replace(\"X\", \"A\")\n",
    "\n",
    "# Remove the first letter from ligand_pdb_seq for 7Y66 and 7XAV - hetatm lines in PDB\n",
    "benchmark_set.loc[benchmark_set[\"pdb\"] == \"7Y66\", \"ligand_pdb_seq\"] = benchmark_set.loc[benchmark_set[\"pdb\"] == \"7Y66\", \"ligand_pdb_seq\"].str[1:]\n",
    "benchmark_set.loc[benchmark_set[\"pdb\"] == \"7XAV\", \"ligand_pdb_seq\"] = benchmark_set.loc[benchmark_set[\"pdb\"] == \"7XAV\", \"ligand_pdb_seq\"].str[1:]\n",
    "\n",
    "# Drop duplicates based on ligand_pdb_seq + name and receptor\n",
    "benchmark_set = benchmark_set.drop_duplicates(subset=[\"ligand_pdb_seq\", \"receptor\"])\n",
    "benchmark_set = benchmark_set.drop_duplicates(subset=[\"ligand_name\", \"receptor\"])\n",
    "\n",
    "# Remove rows where ligand_pdb_sequence is not at least 4 amino acids long\n",
    "benchmark_set = benchmark_set[benchmark_set[\"ligand_pdb_seq\"].str.len() >= 4]\n",
    "\n",
    "# Filter out nearly identical ligands per each receptor\n",
    "filtered_ligands = {}\n",
    "for receptor, group in benchmark_set.groupby('receptor'):\n",
    "    unique_ligands = []\n",
    "    for _, row in group.iterrows():\n",
    "        add_ligand = True\n",
    "        for unique in unique_ligands:\n",
    "            # Calculate similarity between current row's ligand sequence and already added unique ligand sequences\n",
    "            sim = sequence_similarity(row['ligand_pdb_seq'], unique['ligand_pdb_seq'])\n",
    "            if sim > 0.80:\n",
    "                add_ligand = False\n",
    "                break\n",
    "        if add_ligand:\n",
    "            unique_ligands.append(row)\n",
    "    filtered_ligands[receptor] = unique_ligands\n",
    "\n",
    "# Save the benchmark set to a CSV file\n",
    "benchmark_set = pd.concat([pd.DataFrame.from_records([ligand]) for ligands in filtered_ligands.values() for ligand in ligands])\n",
    "benchmark_set.reset_index(drop=True, inplace=True)\n",
    "benchmark_set.to_csv(f\"{outdir}/{outfile}\", index=False)\n",
    "\n",
    "# Save fastas to separate folders\n",
    "os.makedirs(f\"{outdir}/fastas\", exist_ok=True)\n",
    "os.makedirs(f\"{outdir}/fastas/receptors\", exist_ok=True)\n",
    "os.makedirs(f\"{outdir}/fastas/ligands\", exist_ok=True)\n",
    "os.makedirs(f\"{outdir}/fastas/pairs\", exist_ok=True)\n",
    "os.makedirs(f\"{outdir}/fastas/boltz_fastas\", exist_ok=True)\n",
    "os.makedirs(f\"{outdir}/pdbs\", exist_ok=True)\n",
    "os.makedirs(f\"{outdir}/cifs\", exist_ok=True)\n",
    "\n",
    "for index, row in benchmark_set.iterrows():\n",
    "    with open(f\"{outdir}/fastas/receptors/{row['pdb']}_receptor.fasta\", \"w\") as f:\n",
    "        f.write(f\">{row['pdb']}_receptor\\n{row['receptor_pdb_seq']}\")\n",
    "    with open(f\"{outdir}/fastas/ligands/{row['pdb']}_ligand.fasta\", \"w\") as f:\n",
    "        f.write(f\">{row['pdb']}_ligand\\n{row['ligand_pdb_seq']}\")\n",
    "    with open(f\"{outdir}/fastas/pairs/{row['pdb']}.fasta\", \"w\") as f:\n",
    "        f.write(f\">{row['pdb']}_receptor\\n{row['receptor_pdb_seq']}\\n\")\n",
    "        f.write(f\">{row['pdb']}_ligand\\n{row['ligand_pdb_seq']}\\n\")\n",
    "    with open(f\"{outdir}/fastas/boltz_fastas/{row['pdb']}.fasta\", \"w\") as f:\n",
    "        f.write(f\">A|protein\\n{row['receptor_pdb_seq']}\\n\")\n",
    "        f.write(f\">B|protein\\n{row['ligand_pdb_seq']}\\n\")        \n",
    "    # Dowload PDB files\n",
    "    download_structure(row[\"pdb\"], f\"{outdir}/pdbs\")\n",
    "    download_structure(row[\"pdb\"], f\"{outdir}/cifs\", type=\"cif\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
